### ISSUE TITLE
NPC ships occasionally move to (-2400000,-2400000,-2400000)

### ISSUE THREAD

**Issue by smcameron** [smcameron/space-nerds-in-space#147]
Looks like this on the demon screen:

![screenshot from 2018-07-17 09 24 01](https://user-images.githubusercontent.com/23912/42820394-0cef056e-89a4-11e8-9077-c193afaec784.png)

I don't know why they do this.

Coordinates come into snis_client in process_update_econ_ship_packet as -2400000, -2400000, -2400000, but (so far as I can tell) do not have these values in snis_server in send_econ_update_ship_packet().

So, possibly a marshaling problem?

The buffer received by the client in process_update_econ_ship_packet looks like this:

```
00 00 02 59 00 00 01 51 00 01 80 00 00 00 80 00 00 00 80 00 00 00 00 00 00 00 00 00 00 00 ff ff ff ff 14 03 ff
```

**Comment by smcameron** [comment]
So the format of the buffer is "wwhSSSQwbbb", wih the SSS being the x,y,z coordinates, and those bytes all have the same values in the buffer of:

```
80 00 00 00
```

which 0x80 0x00 0x00 0x00 as "S" decodes as -2400000.001118

**Comment by smcameron** [comment]
Ok, I caught the server sending bad data... the ship coords are NaNs. :(

```
#3  0x0000000000469285 in send_econ_update_ship_packet (c=0x872580 <client>, o=0xcd0d00 <go+1367040>)
    at snis_server.c:19802
19802		pb_queue_to_client(c, packed_buffer_new("bwwhSSSQwbbb", opcode,
(gdb) print o->x
$1 = -nan(0x8000000000000)
(gdb) print o->y
$2 = -nan(0x8000000000000)
(gdb) print o->z
$3 = -nan(0x8000000000000)
(gdb)
```

**Comment by smcameron** [comment]
Ok, I think the NaNs causing this were coming from temper_velocity_change() when a ship was trying to change it's velocity to be 180 degrees from current velocity (i.e. reverse). The cross product ended up zero, and then we tried to normalize the cross product, which got us NaNs.

This fixes it: b9c1b3faa2685dd3fe7ccc25fd500bb1fbdd5b69

Although I'm not sure that's the best fix (maybe for reversing, I should modify the desired velocity to something like 45 degrees off current velocity?)

Also, I am not certain there are no other NaNs lurking around, but this one was a major source of them.

**Comment by smcameron** [comment]
Fixed another source of NaNs with 4e3a0f5403b9e65604a37cb5117b3fb2c608f766

Not aware of any other sources of NaNs.
